### ISSUE TITLE
[ORM] Inconsistent DB ORM casting of int values

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-framework#7039]
This covers a range of issues, but is essentially caused by the lack of consistency between the various DB connector classes in the way that they encode integer values. Some encode them as strings, others as real integer.

This issue causes increasing breakage in client scripts where we work with de-serialised client side models, but in a type-strict way. E.g. `this.props.id === checkID`, which may not have consistent behaviour when using another DB backend.

I won't propose a desired resolution, but previously discussed suggestions included:
 - Enforce strictly-typed DB connectors (e.g. PDO, move to DBAL (https://github.com/silverstripe/silverstripe-framework/issues/6632))
 - Cast int fields at the DataObject level (cast on write, store normalised values)
 - Cast values during serialisation for the front-end
 
 This will also fix:

 * #3473
 * #6709

# Related Pull Requests
* [x] https://github.com/silverstripe/silverstripe-postgresql/pull/91
* [ ] https://github.com/silverstripe/silverstripe-framework/pull/8448

**Comment by chillu** [comment]
@tractorcow Can you explain how this is an `impact/high`? It seems more like a minor nuisance that gets a bit more attention now that we do type strict comparisons on more serialised frontend data? Which is a particular use case we could cover through the GraphQL layer. "Move to DBAL" definitely won't happen for 4.0. What feasible measure do you propose for this milestone? I'm wondering if we lower impact and remove from milestone.

**Comment by tractorcow** [comment]
The impact/high would be the impact on all type-safe checks that would be affected should we update this behaviour. It would potentially break checks which may rely on legacy typing. E.g. validation of form submission strings vs int values from the database.

**Comment by tractorcow** [comment]
Turned out to be not as hard as I had feared; Although the solution I've given only works with mysqlnd.

**Comment by tractorcow** [comment]
I propose to close this with the following justification:

 - There are few reliable ways to enforce this at the DB connector level.
 - Enforcing this at the model level would require additional overhead / performance considerations.
 - Enforcing this internally would be akin to attempting to "fix" php dynamic typing, and would essentially be making mistyping illegal within the ORM. Given that PHP is inherently dynamically typed, this would contradict standard PHP development practices.
 
In short, dynamic typing is a part of PHP, we shouldn't be trying to fight it because we dislike it.

**Comment by tractorcow** [comment]
Proof of concept of the weakness of EMULATE_PREPARES = false https://gist.github.com/tractorcow/390c9075020b6e2fdfa807c835403afa

**Comment by sminnee** [comment]
> Enforcing this internally would be akin to attempting to "fix" php dynamic typing, and would essentially be making mistyping illegal within the ORM. Given that PHP is inherently dynamically typed, this would contradict standard PHP development practices.

I disagree with this, but it might be okay for it to be SS5 thing with a minimum requirement of PHP 7.1+. The type-hygiene of PHP, JavaScript and other languages is on a general improving trajectory.

When SS2 was first developed, you could have made a similar argument about OOP. Thankfully, SS isn't a function soup otherwise I don't think we'd be using it today. ;)

> Enforcing this at the model level would require additional overhead / performance considerations.

I'd rather not slow down things further but also I don't know what the impact of this would be.

**Comment by tractorcow** [comment]
Yes, I would agree with this if we moved to php7 min with `declare(strict_types=1);` enabled. Otherwise it would be an exercise in futility.

**Comment by tractorcow** [comment]
I'm moving this out of v4 milestone for now.

**Comment by sminnee** [comment]
Picking this up again.

I think we should reconsider the approach recommended in #7337 and make mysqlnd / mysqli the recommended versions to use (maybe throw a warning in the installer or somewhere if these aren't available). The other plugin is generally listed as deprecated; mysqlnd / mysqli was recommended as the default since php 5.4.

We should check the status of PGSQL and maybe SQLite's type safety, but I suspect that they're better than MySQL at this.

Trying to apply band-aids over badly-written database drivers will make the ORM slower and I don't think it's necessary. SilverStripe will still work if the types aren't clean, but there will be a few minor issues like #6709.

**Comment by sminnee** [comment]
Still got the framework PR to merge before this can be closed!
